src/reference-manual/compiler-optimizations.Rmd

```diff
@@ -0,0 +1,504 @@
+
+Example Stan program:
+
+```
+data {
+  int m;
+}
+transformed data {
+  int n = m+1;
+  int a[n];
+  for (i in 1:n-1) {
+    a[i] = i;
+  }
+}
+```
+
+Compiler representation of program **before expression propagation** (simplified from the output of `--debug-transformed-mir-pretty`):
+
+```
+prepare_data {
+  data int m = FnReadData__("m")[1];
+  data int n = (m + 1);
+  data array[int, n] a;
+  for(i in 1:(n - 1)) {
+    a[i] = i;
+  }
+}
+```
+
+Compiler representation of program **after expression propagation** (simplified from the output of `--debug-optimized-mir-pretty`):
+
+```
+prepare_data {
+  data int m = FnReadData__("m")[1];
+  data int n = (m + 1);
+  data array[int, (m + 1)] a;
+  for(i in 1:((m + 1) - 1)) {
+    a[i] = i;
+  }
+}
+```
+
+
+### Copy propagation
+
+Copy propagation is similar to [expression propagation](#org4e2cadd), but only propagates variables rather than arbitrary expressions. This can reduce the complexity of the code for other optimizations such as expression propagation.
+
+Example Stan program:
+
+```
+model {
+  int i = 1;
+  int j = i;
+  int k = i + j;
+}
+```
+
+Compiler representation of program **before copy propagation** (simplified from the output of `--debug-transformed-mir-pretty`):
+
+```
+log_prob {
+    int i = 1;
+    int j = i;
+    int k = (i + j);
+}
+```
+
+Compiler representation of program **after copy propagation** (simplified from the output of `--debug-optimized-mir-pretty`):
+
+```
+log_prob {
+  int i = 1;
+  int j = i;
+  int k = (i + i);
+}
+```
+
+
+### Partial evaluation
+
+Partial evaluation searches for expressions that can be replaced with a faster, simpler, more memory efficient, or more numerically stable expression that has the same meaning.
+
+Example Stan program:
+
+```
+model {
+  real a = 1 + 1;
+  real b = log(1 - a);
+  real c = a + b * 5;
+}
+```
+
+Compiler representation of program **before partial evaluation** (simplified from the output of `--debug-transformed-mir-pretty`):
+
+```
+log_prob {
+  real a = (1 + 1);
+  real b = log((1 - a));
+  real c = (a + (b * 5));
+}
+```
+
+Compiler representation of program **after partial evaluation** (simplified from the output of `--debug-optimized-mir-pretty`):
+
+```
+log_prob {
+  real a = 2;
+  real b = log1m(a);
+  real c = fma(b, 5, a);
+}
+```
+
+
+### Lazy code motion
+
+<a id="org0c605fa"></a> Lazy code motion rearranges the statements and expressions in a program with the goals of:
+
+-   Avoiding computing expressions more than once, and
+-   Computing expressions as late as possible (to minimize the strain on the working memory set).
+
+To accomplish these goals, lazy code motion will perform optimizations such as:
+
+-   Moving a repeatedly calculated expression its own variable (also referred to as *common-subexpression elimination*)
```

Comment:
> expression its own variable

"expression to its own variable"

I'd also say Assigning rather than Moving, but if Moving is the appropriate lingo stick with that.

Reply:
I think moving is the right word here. If we said assignment I think we would have to say something like `Assign a repeatedly calculated expression a temporary'